zeebe/gateway-rest/src/test/java/io/camunda/zeebe/gateway/rest/controller/UserControllerTest.java

```diff
@@ -0,0 +1,290 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.zeebe.gateway.rest.controller;
+
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import io.camunda.identity.usermanagement.CamundaUser;
+import io.camunda.identity.usermanagement.CamundaUserWithPassword;
+import io.camunda.identity.usermanagement.service.UserService;
+import io.camunda.zeebe.gateway.protocol.rest.CamundaUserDto;
+import io.camunda.zeebe.gateway.protocol.rest.CamundaUserWithPasswordDto;
+import io.camunda.zeebe.gateway.protocol.rest.SearchRequestDto;
+import io.camunda.zeebe.gateway.protocol.rest.UserSearchResponseDto;
+import io.camunda.zeebe.gateway.rest.GlobalControllerExceptionHandler;
+import io.camunda.zeebe.gateway.rest.controller.UserControllerTest.TestUserControllerApplication;
+import io.camunda.zeebe.gateway.rest.controller.usermanagement.UserController;
+import java.net.URI;
+import java.util.List;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.boot.autoconfigure.jdbc.DataSourceAutoConfiguration;
+import org.springframework.boot.autoconfigure.orm.jpa.HibernateJpaAutoConfiguration;
+import org.springframework.boot.autoconfigure.security.servlet.SecurityAutoConfiguration;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.test.context.SpringBootTest.WebEnvironment;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.http.ProblemDetail;
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.test.web.reactive.server.WebTestClient;
+
+@SpringBootTest(
+    classes = {
+      TestUserControllerApplication.class,
+      UserController.class,
+      GlobalControllerExceptionHandler.class
+    },
+    webEnvironment = WebEnvironment.RANDOM_PORT)
+public class UserControllerTest {
```

Comment:
🔧 Please have a look at the other controller tests, e.g. the `UserTaskControllerTest`. They are simple `@WebMvcTest`s instead of `@SpringBootTest`s. That way, you can scope the test class to exactly one controller and all its required dependencies instead of having to mock all potential beans that might be created along the way. Otherwise, we'll have to adjust this test for EACH new bean we create that will automatically be needed here as well since we start a Spring application with the full application context.

In the WebMvc tests, you'll also rather use String input for calls instead of using defined DTOs generated from OpenAPI. This helps verifying that such calls actually (de-)serialize correctly AND it's simpler to test that data was NOT put into a request so testing missing/malformed input is easier as well.

Reply:
Thanks, that makes absolute sense! Great learning for me 🙏 🥳 

I'll adapt it accordingly right away...